engine/src/main/java/io/camunda/zeebe/engine/state/compensation/DbCompensationSubscriptionState.java

```diff
@@ -0,0 +1,79 @@
+
+import io.camunda.zeebe.db.ColumnFamily;
+import io.camunda.zeebe.db.TransactionContext;
+import io.camunda.zeebe.db.ZeebeDb;
+import io.camunda.zeebe.db.impl.DbCompositeKey;
+import io.camunda.zeebe.db.impl.DbLong;
+import io.camunda.zeebe.db.impl.DbString;
+import io.camunda.zeebe.db.impl.DbTenantAwareKey;
+import io.camunda.zeebe.db.impl.DbTenantAwareKey.PlacementType;
+import io.camunda.zeebe.engine.state.mutable.MutableCompensationSubscriptionState;
+import io.camunda.zeebe.protocol.ZbColumnFamilies;
+import io.camunda.zeebe.protocol.impl.record.value.compensation.CompensationSubscriptionRecord;
+
+public class DbCompensationSubscriptionState implements MutableCompensationSubscriptionState {
+
+  private final DbLong processInstanceKey;
+  private final DbString compensableActivityIdKey;
+  private final DbString tenantIdKey;
+  private final DbTenantAwareKey<DbLong> tenantAwareProcessInstanceKey;
+  private final DbCompositeKey<DbTenantAwareKey<DbLong>, DbString>
+      tenantAwareProcessInstanceKeyCompensableActivityId;
+  private final ColumnFamily<
+          DbCompositeKey<DbTenantAwareKey<DbLong>, DbString>, CompensationSubscription>
+      compensationSubscriptionColumnFamily;
+  private final CompensationSubscription compensationSubscription = new CompensationSubscription();
+
+  public DbCompensationSubscriptionState(
+      final ZeebeDb<ZbColumnFamilies> zeebeDb, final TransactionContext transactionContext) {
+    processInstanceKey = new DbLong();
+    compensableActivityIdKey = new DbString();
+    tenantIdKey = new DbString();
+    tenantAwareProcessInstanceKey =
+        new DbTenantAwareKey<>(tenantIdKey, processInstanceKey, PlacementType.PREFIX);
+    tenantAwareProcessInstanceKeyCompensableActivityId =
+        new DbCompositeKey<>(tenantAwareProcessInstanceKey, compensableActivityIdKey);
+    compensationSubscriptionColumnFamily =
+        zeebeDb.createColumnFamily(
+            ZbColumnFamilies.COMPENSATION_SUBSCRIPTION,
+            transactionContext,
+            tenantAwareProcessInstanceKeyCompensableActivityId,
+            compensationSubscription);
+  }
+
+  @Override
+  public CompensationSubscription get(
+      final String tenantId, final long processInstanceKey, final String compensableActivityId) {
+    wrapCompensationKeys(processInstanceKey, compensableActivityId, tenantId);
+    return compensationSubscriptionColumnFamily.get(
+        tenantAwareProcessInstanceKeyCompensableActivityId);
+  }
+
+  @Override
+  public void put(final long key, final CompensationSubscriptionRecord compensation) {
+    compensationSubscription.setKey(key).setRecord(compensation);
+
+    wrapCompensationKeys(
+        compensation.getProcessInstanceKey(),
+        compensation.getCompensableActivityId(),
+        compensation.getTenantId());
+
+    compensationSubscriptionColumnFamily.upsert(
+        tenantAwareProcessInstanceKeyCompensableActivityId, compensationSubscription);
+  }
```

Comment:
Can you clarify this ? Are you talking about using `upsert` to store the compensation subscription ? Or we should add a check to verify if the activityId is already present in the state ?

Reply:
To implement https://github.com/camunda/zeebe/issues/15141 or https://github.com/camunda/zeebe/issues/15152, we will need to store multiple subscriptions per element. Since the subscriptions will have the same element id, we will need to store the subscriptions differently in the state, for example by a composite key of `<tenant-id, process-instance-key, subscription-key>`.

Reply:
Ok, thanks for the explanation